### ISSUE TITLE
Error in middleware/secondary/setup.go: no f in case "transfer"

### ISSUE THREAD

**Issue by snowirbis** [coredns/coredns#716]
This error prevents secondary zones recognized as "true" secondary, so secondary setting never worked.

```
diff `--git a/middleware/secondary/setup.go b/middleware/secondary/setup.go
index ca665e7..64215d0 100644
--- a/middleware/secondary/setup.go
+++ b/middleware/secondary/setup.go
@@ -70,7 +70,7 @@ func secondaryParse(c *caddy.Controller) (file.Zones, error) {

                                switch c.Val() {
                                case "transfer":
-                                       t, _, e = file.TransferParse(c, true)
+                                       t, f, e = file.TransferParse(c, true) // just add `f`
                                        if e != nil {
                                                return file.Zones{}, e
                                        }

```

**Comment by miekg** [comment]
PR (+ test showing the problem!) welcome :)

**Comment by snowirbis** [comment]
Sorry, Miek, I'm very beginner in Go and didn't learned to write Go tests yet. 
I'm using this commit 36396e94ab738439aa9cfc0fe9ed5db2fca883d3 and found this bug or maybe typo. My config is really simple:

mazur.com.ua:53 {
  secondary {
    transfer from 193.151.xx.xxx
    transfer from 91.194.xxx.xx
  }
  errors stdout
  log stdout
}

When I starts CoreDNS, without a patch (I was previously submitted), it starts, but cant' transfer zone from masters. This is happening because `f ` not returned (as `from`) in line 73 in coredns/middleware/secondary/setup.go, so I submitted this little patch.

Thanks for you attention, youre doing a great work!
Wish you all the best

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Error in midd..." ]
>Sorry, Miek, I'm very beginner in Go and didn't learned to write Go tests yet.
>I'm using this commit 36396e94ab738439aa9cfc0fe9ed5db2fca883d3 and found this bug or maybe typo. My config is really simple:
>
>mazur.com.ua:53 {
>  secondary {
>    transfer from 193.151.xx.xxx
>    transfer from 91.194.xxx.xx
>  }
>  errors stdout
>  log stdout
>}
>
>When I starts CoreDNS, without a patch (I was previously submitted), it starts, but cant' transfer zone from masters. This is happening because `f ` not returned (as `from`) in line 73 in coredns/middleware/secondary/setup.go, so I submitted this little patch.
>
>Thanks for you attention, youre doing a great work!
>Wish you all the best

Thanks and no worries.

Do you some logging? What do you see, it just does not pull the zone?

**Comment by snowirbis** [comment]
Yep, just doesn't recognize zone as a secondary, because no IP readed from `from`.
I done some researching on `TransferFrom` and locate `_`  where `f` should be.
Already done a PR. 

I'm newbie in Go and new to GitHub features, but I'm perl old-schooler.

Have a nice evening!

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Error in midd..." ]
>Yep, just doesn't recognize zone as a secondary, because no IP readed from `from`.
>I done some researching on `TransferFrom` and locate `_`  where `f` should be.
>Already done a PR.
>
>I'm newbie in Go and new to GitHub features, but I'm perl old-schooler.
>
>Have a nice evening!

Great. I saw your PR, but I also want to repro it, so I can add a test.

(Old perl fan myself)

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
Ah, I think I see, if you want to transfer the zone *again*, you'll need an explicit `transfer to *` statement in your config:

~~~
<zone>.org {
    secondary {
        transfer to *
        transfer from 82.161.xxx.xxx
    }
    import /etc/coredns/zone-default
}
~~~
Is a setup I have.

PR to update the documentation to make this clearer welcome.

**Comment by snowirbis** [comment]
Not "again", zone transfer should start on coredns launch, but id does not.

**Comment by miekg** [comment]
Ok. But you'll need to be more specific on what works, what doesn't work,
what config you are using and in the end even tcpdumps. Because I can't
reproduce and I seem to be using the exact same setup as you.

**Comment by snowirbis** [comment]
Okay, but no dump available, because no packets going out, zone does not transferred, because 'secondary' middleware not recognize configured xone as a secondary one. I will do a video record of what I exactly do and how to fix it, and what about my 'diff' is :) But for now I must have some rest.

Thanks and have a nice evening!

**Comment by miekg** [comment]
That's OK. But `dig axfr <zone> @remote-ip` works? But CoreDNS does not transfer?

**Comment by snowirbis** [comment]
Exactly! 
`axfr @remote mazur.com.ua` works as expected. 

Just two screenshots:

Before editing of coredns/coredns/middleware/secondary/setup.go: https://prnt.sc/fi5c6c
And after: (`f` now on it's place) http://prntscr.com/fi5ct1

Just to clarify my DNS skills: I'm local domain names registrar in .ua for about 15 years, and clearly understand what exactly I'm doing with all the DNS stuff :)

**Comment by miekg** [comment]
Ack :) can you email me the exact details? I just have to see this for
myself. Thanks!

**Comment by snowirbis** [comment]
Of course i do, in a couple of hours.

**Comment by miekg** [comment]
Did you send those details? (Haven't seen anything)

**Comment by SuperQ** [comment]
I can also confirm this, with current coredns (>= 007), secondary does not setup correctly and my primary server sees no transfer queries.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Error in midd..." ]
>I can also confirm this, with current coredns (>= 007), secondary does not setup correctly and my primary server sees no transfer queries.

ah good :) Should be able to extract a testcase. But I'm still confused what the 
sympton is. Who axfring from where?
(I'm using the same setup (i think) and everything is working...)

**Comment by SuperQ** [comment]
I have a setup like this:
`ns.example.com` (bind9, master for zones)
`ns2.example.com` (coredns, secondary only)

My corefile looks like this:
```
example.com {
  prometheus
  log stdout { class error }
  secondary {
    transfer from 1.2.3.4
    transfer from 2602::1
  }
}
```

**Comment by SuperQ** [comment]
If I trigger a notify event on my bind9, with coredns 008 secondary:

```
2017/06/19 11:45:00 [INFO] Dropping notify from 1.2.3.4 for example.com.
```

**Comment by miekg** [comment]
Thanks for that snippets - trying to replicate locally by running 3 CoreDNSs and thinks definitely look fishy.

**Comment by miekg** [comment]
@snowirbis you were right, sorry for not jumping on this sooner. There are some other bits that concern me though (while playing with this). And we obviously have no test for it

**Comment by snowirbis** [comment]
Hi Miek! 
I dind't send you detailed video report due to lack of time, sorry. This little hardcatchable bug makes me crazy :) Thanks for finding and fixing it!

**Comment by snowirbis** [comment]
Here is https://github.com/coredns/coredns/pull/745/commits/45c874c5eff6b06b275b69dff47cf5739a3d62a9 exactly what I mean in https://github.com/coredns/coredns/issues/716#issue-234507251 :)

**Comment by miekg** [comment]
yes, your patch was good. I now however also discovered a data race (sort of expected), while making the test and fix a bunch of other things, see https://github.com/coredns/coredns/pull/745

I can still merge your fix though - which would be good, as it is *your* fix.

**Comment by snowirbis** [comment]
thanks. yes, already saw 745, that's great!

**Comment by miekg** [comment]
Fixed with #745
